main/src/library/MutDisjointSet.flix

```diff
@@ -0,0 +1,149 @@
+            MutMap.put!(x, singleton(rc, x), forest)
+        } else {
+            ()
+        }
+    }
+
+    ///
+    /// Returns `true` iff `s` contains `x`.
+    ///
+    pub def contains(x: t, s: MutDisjointSet[t, r]): Bool \ r with Order[t] = {
+        let MutDisjointSet({ forest | _ }) = s;
+        match MutMap.get(x, forest) {
+            case Some(_) => true
+            case None    => false
+        }
+    }
+
+    ///
+    /// Returns a new `Node` from `x` in `rc`.
+    ///
+    def singleton(rc: Region[r], x: t): Node[t, r] \ r =
+        Node.Node({ rc = rc, value = x, parent = Ref.new(rc, None), rank = Ref.new(rc, 0) })
+
+    ///
+    /// Updates `s` with collection of disjoint singleton sets containing
+    /// the elements of the iterable `m`.
+    ///
+    pub def makeSets(m: m[t], s: MutDisjointSet[t, r]): Unit \ r with Iterable[m], Order[t] = region rc {
+        Iterable.iterator(rc, m) |> Iterator.forEach(flip(makeSet)(s))
+    }
+
+    ///
+    /// Returns the representative / root of the set that contains `x`.
+    ///
+    pub def find(x: t, s: MutDisjointSet[t, r]): Option[t] \ r with Order[t] =
+        findSet(x, s) |> Option.map(match Node.Node({ value | _ }) -> value)
+
+    ///
+    /// Returns the representative / root `Node` of the set that contains `x`.
+    ///
+    def findSet(x: t, s: MutDisjointSet[t, r]): Option[Node[t, r]] \ r with Order[t] = {
+        def findRoot(y) = {
+            let Node.Node({ parent | _ }) = y;
+            match Ref.get(parent) {
+                case Some(p) => {
+                    let root = findRoot(p);
+                    Ref.put(Some(root), parent);
+                    root
+                }
+                case None => y
+            }
+        };
+        let MutDisjointSet({ forest | _ }) = s;
+        MutMap.get(x, forest)
+        |> Option.map(findRoot)
+    }
+
+    ///
+    /// Merges the sets that contain `x` and `y` if both `x` and `y` are in `s`.
+    /// If `x` and `y` have the same rank, then `y` becomes the new parent of `x`.
+    ///
+    pub def union(x: t, y: t, s: MutDisjointSet[t, r]): Unit \ r with Order[t] =
+        def link(x1, y1) = {
+            if (not equal(x1, y1)) {
+                let Node.Node({ parent = xparent, rank = xr | _ }) = x1;
+                let Node.Node({ parent = yparent, rank = yr | _ }) = y1;
+                let xrank = Ref.get(xr);
+                let yrank = Ref.get(yr);
+                if (xrank > yrank) {
+                    Ref.put(Some(x1), yparent)
+                } else {
+                    Ref.put(Some(y1), xparent)
+                };
+                if (xrank == yrank) {
+                    Ref.put(yrank + 1, yr)
+                } else {
+                    ()
+                }
+            }
+            else {
+                ()
+            }
+        };
+        match (findSet(x, s), findSet(y, s)) {
+            case (Some(xn), Some(yn)) => link(xn, yn)
+            case _ => ()
```

Comment:
So if we have forgotten to call makeSet this does not work so great or?

Reply:
Yes, then it does nothing. I think that is easier to understand than adding the element to the set. What do you think? I will update the documentation to reflect whatever decision we make :)

Reply:
I think doing nothing is dangerous. What would be better?

Reply:
A crash or some other behavior seems preferable.

Reply:
Any reason we could not start by adding x and y to the forest if they are not there?

Reply:
> I think doing nothing is dangerous. What would be better?

How so?

I don't think it's the responsibility of `union` to do this.
I think it *feels* weird, that `union` has this extra side effect that effectively allows you to do what `makeSet` does. This assumes that the caller knows what elements they are working with, and prevents them from accidentally growing the forest. If they want that, they can do so explicitly by first calling `makeSet` (recall that it doesn't add elements that already exist) then `union`.
In a sense, it's more predictable; change the structure if the structure contains the elements, otherwise do nothing.

Reply:
We could also completely drop makeSets? What's its purpose with _this_ specific API?

It seems that the only purpose is that you must perform some action before calling union. If you don't then things go wrong. But does it have to be so? We could just say that elements we have not seen before (implicitly) belong to singleton sets. 

Imagine you want disjoint elements of integers, but you don't know all of them upfront. Now calling makeSets is cumbersome.

Reply:
I think we still need `makeSet(s)` as it is a canonical function, but I agree that we can make the union operation add new elements.

Reply:
I think there are two designs: __either__ you need `makeSet` and you work with `Node`s __or__ you never work with `Node`s and you don't need `makeSet`. I think we are mixing them. So I would propose we drop `makeSet`. This is potentially a bit slower, but it makes the API much more flexible and clear.

Reply:
Does it make sense?

Reply:
If we drop `makeSet`, it seems that we have an implicit assumption that **every** element of a type is in a singleton set when making a `MutDisjointSets` of that type. Thus, only unions are stored. However, this may not always be what a user wants. Sometimes, they may want to query the data structure on a finite subset (/ finite amount of elements) of the type. That's why I think we need both.

Otherwise, you could add singleton sets to the forest by performing a union with the same element. That seems counterintuitive, though.

Reply:
> If we drop makeSet, it seems that we have an implicit assumption that every element of a type is in a singleton set when making a MutDisjointSets of that type.

Yes. That seems reasonable no? Like `List[Int32]` works for any `Int32`.

Consider an application: Type inference. Here link type variables with other type variables. But type inference may also create new variables on-the-fly. Having to call makeSet upfront is now impossible. Having to call it before link is cumbersome.

What alternative do you propose? And how does it avoid either silent wrong behavior or crashing the program?

Reply:
> Yes. That seems reasonable no? Like `List[Int32]` works for any `Int32`.

You still have to add each element to the list. They are not all implicitly part of the list.

> Having to call makeSet upfront is now impossible. Having to call it before link is cumbersome.

I would say it's not cumbersome but explicit.

> And how does it avoid either silent wrong behavior or crashing the program?

Returning a `Result` or `Option`? It seems to me that allowing "unioning" two elements that are not explicitly stored in the data structure could also be viewed as silent wrong behavior. Ultimately, it's just a matter of what we put in the documentation.

> I think there are two designs: either you need makeSet and you work with Nodes or you never work with Nodes and you don't need makeSet. I think we are mixing them. So I would propose we drop makeSet. This is potentially a bit slower, but it makes the API much more flexible and clear.

I think we still need `Node` to store rank and parent pointers.



I think the API you are suggesting is totally fine, but I just wanted to raise some of my concerns / opinions before moving on with it :)